include/boost/beast/core/impl/buffers_adaptor.hpp

```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+subrange(
+    iter_type first,  // iterator to first buffer in storage
+    iter_type last,   // iterator to last buffer in storage
+    std::size_t pos,        // the offset in bytes from the beginning of the storage
+    std::size_t n)         // the total length of the subrange
+    : first_(first)
+    , last_(last)
+    , first_offset_(0)
+    , last_size_((std::numeric_limits<std::size_t>::max)())
+{
+    adjust(pos, n);
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+adjust(
+    std::size_t pos,
+    std::size_t n)
+-> void
+{
+    if (n == 0)
+        last_ = first_;
+
+    // note: fallthrough
+
+    if (first_ == last_)
+    {
+        first_offset_ = 0;
+        last_size_ = 0;
+        return;
+    }
+
+    auto last_element = [this](iter_type iter) {
+        return std::next(iter) == last_;
+    };
+
+
+    pos += first_offset_;
+    while (pos)
+    {
+        auto adjust = (std::min)(pos, first_->size());
+        if (adjust >= first_->size())
+        {
+            ++first_;
+            first_offset_ = 0;
+            pos -= adjust;
+        }
+        else
+        {
+            first_offset_ = adjust;
+            pos = 0;
+        }
+    }
+
+    auto current = first_;
+    auto max_elem = current->size() - first_offset_;
+    if (last_element(current))
+    {
+        // both first and last element
+        last_size_ = (std::min)(max_elem, n);
+        last_ = std::next(current);
+        return;
+    }
+    else if (max_elem >= n)
+    {
+        last_ = std::next(current);
+        last_size_ = n;
+    }
+    else
+    {
+        n -= max_elem;
+        ++current;
+    }
+
+    for (;;)
+    {
+        max_elem = current->size();
+        if (last_element(current))
+        {
+            last_size_ = (std::min)(n, last_size_);
+            return;
+        }
+        else if (max_elem < n)
+        {
+            n -= max_elem;
+            ++current;
+        }
+        else
+        {
+            last_size_ = n;
+            last_ = std::next(current);
+            return;
+        }
+    }
+}
+
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+begin() const -> iterator
+{
+    return iterator(this, first_);
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+end() const -> iterator
+{
+    return iterator(this, last_);
+}
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+get_buffer(iter_type it) const
+-> value_type
+{
+    auto result = *it;
+
+    if (it == first_)
+        result += first_offset_;
+
+    if (std::next(it) == last_)
+        detail::truncate(result, last_size_);
```

Comment:
There's no need to have a separate `truncate` function, since it is only called from here. And even if it wasn't, the operation is so trivial that it is not deserving of being elevated to its own function.

Reply:
strongly disagree.
My reasoning is that one logical operation deserves its own function and name as this aids higher level reasoning about intent.
Again, will change.